File: go/lexer_test.go
package lexer

import (
	"testing"
)

func TestLexer(t *testing.T) {
	testCases := []struct {
		name     string
		input    string
		expected []Token
	}{
		{
			name:  "Simple Tokens",
			input: "=+(){},;",
			expected: []Token{
				NewToken(Equal, "="),
				NewToken(Plus, "+"),
				NewToken(LParen, "("),
				NewToken(RParen, ")"),
				NewToken(LSquirly, "{"),
				NewToken(RSquirly, "}"),
				NewToken(Comma, ","),
				NewToken(Semicolon, ";"),
				NewToken(Eof, ""),
			},
		},
		{
			name: "Complete Token",
			input: `let five = 5;
				let ten = 10;
				let add = fn(x, y) {
					x + y;
				};
				let result = add(five, ten);`,
			expected: []Token{
				NewToken(Let, "let"),
				NewToken(Ident, "five"),
				NewToken(Equal, "="),
				NewToken(Int, "5"),
				NewToken(Semicolon, ";"),
				NewToken(Let, "let"),
				NewToken(Ident, "ten"),
				NewToken(Equal, "="),
				NewToken(Int, "10"),
				NewToken(Semicolon, ";"),
				NewToken(Let, "let"),
				NewToken(Ident, "add"),
				NewToken(Equal, "="),
				NewToken(Function, "fn"),
				NewToken(LParen, "("),
				NewToken(Ident, "x"),
				NewToken(Comma, ","),
				NewToken(Ident, "y"),
				NewToken(RParen, ")"),
				NewToken(LSquirly, "{"),
				NewToken(Ident, "x"),
				NewToken(Plus, "+"),
				NewToken(Ident, "y"),
				NewToken(Semicolon, ";"),
				NewToken(RSquirly, "}"),
				NewToken(Semicolon, ";"),
				NewToken(Let, "let"),
				NewToken(Ident, "result"),
				NewToken(Equal, "="),
				NewToken(Ident, "add"),
				NewToken(LParen, "("),
				NewToken(Ident, "five"),
				NewToken(Comma, ","),
				NewToken(Ident, "ten"),
				NewToken(RParen, ")"),
				NewToken(Semicolon, ";"),
				NewToken(Eof, ""),
			},
		},
	}

	for _, tc := range testCases {
		t.Run(tc.name, func(t *testing.T) {
			l := NewLexer(tc.input)

			for i, expectedToken := range tc.expected {
				token := l.NextToken()
				checkToken(t, token, expectedToken, i)
			}
		})
	}
}

func checkToken(t *testing.T, token, expectedToken Token, index int) {
	if token.Type != expectedToken.Type {
		t.Errorf("Token %d - tokentype wrong. expected=%q, got=%q", index, expectedToken.Type, token.Type)
	}
	if token.Literal != expectedToken.Literal {
		t.Errorf("Token %d - literal wrong. expected=%q, got=%q", index, expectedToken.Literal, token.Literal)
	}
}

File: scala/lexer.test.scala
//> using test.dep org.scalatest::scalatest::3.2.16

import org.scalatest.wordspec.AnyWordSpec
import org.scalatest.matchers.must
import org.scalatest.prop.TableDrivenPropertyChecks._

class LexerSpec extends AnyWordSpec with must.Matchers {

  "parseTokens()" must {

    "throw on unexpected input that is not currently handled/implemented" in {
      assertThrows[AssertionError](parseTokens("\uD83E\uDD5C\uD83D\uDE33"))
    }

    "be able to parse a single char token" in {
      val cases = Table(
        ("Input", "Expected result"),
        ("=", Token.Equal),
        ("+", Token.Plus),
        (",", Token.Comma),
        (";", Token.Semicolon),
        ("{", Token.LSquirly),
        ("}", Token.RSquirly),
        ("(", Token.LParen),
        (")", Token.RParen),
        (" ", Token.Whitespace),
        ("\t", Token.Whitespace)
      )

      forAll(cases) { (input: String, expected: Token) =>
        parseTokens(input) must be(Seq(expected))
      }
    }

    "be able to parse a sequence of char tokens" in {
      parseTokens("=+(){},") must be(
        Seq(
          Token.Equal,
          Token.Plus,
          Token.LParen,
          Token.RParen,
          Token.LSquirly,
          Token.RSquirly,
          Token.Comma
        )
      )
    }

    "be able to parse a keyword token" in {
      parseTokens("let") must be(Seq(Token.Let))
      parseTokens("fn") must be(Seq(Token.Function))
    }

    "be able to parse a string literal" in {
      parseTokens("\"hello\"") must be(Seq(Token.Str("hello")))
    }

    "be able to parse an integer literal" in {
      parseTokens("123") must be(Seq(Token.Integer(123)))
    }

    "produce Token.Invalid for a number that exceeds the range of an int" in {
      val input = Integer.MAX_VALUE.toString + "0"
      parseTokens(input) must be(Seq(Token.Invalid))
    }

    "be able to parse an identifier starting with an underscore" in {
      parseTokens("_hello") must be(Seq(Token.Identifier("_hello")))
    }

    "be able to parse an identifier with a number" in {
      parseTokens("hello123") must be(Seq(Token.Identifier("hello123")))
    }

    "be able to parse a simple assignment" in {
      parseTokens("let x = 5;").filterNot(_ == Token.Whitespace) must be(
        Seq(
          Token.Let,
          Token.Identifier("x"),
          Token.Equal,
          Token.Integer(5),
          Token.Semicolon
        )
      )
    }

    "be able to parse complex input" in {
      val input: String =
        """
          |let five = 5;
          |let ten = 10;
          |
          |let add = fn(x, y) {
          |   x + y;
          |};
          |
          |let result = add(five, ten);
          |""".stripMargin

      // Don't care about whitespace for assertions, discard it after parsing
      val res = parseTokens(input).filterNot(_ == Token.Whitespace)

      res must be(
        Seq(
          // let five = 5;
          Token.Let,
          Token.Identifier("five"),
          Token.Equal,
          Token.Integer(5),
          Token.Semicolon,

          // let ten = 10;
          Token.Let,
          Token.Identifier("ten"),
          Token.Equal,
          Token.Integer(10),
          Token.Semicolon,

          // let add = fn(x, y) {
          //   x + y;
          // };
          Token.Let,
          Token.Identifier("add"),
          Token.Equal,
          Token.Function,
          Token.LParen,
          Token.Identifier("x"),
          Token.Comma,
          Token.Identifier("y"),
          Token.RParen,
          Token.LSquirly,
          Token.Identifier("x"),
          Token.Plus,
          Token.Identifier("y"),
          Token.Semicolon,
          Token.RSquirly,
          Token.Semicolon,

          // let result = add(five, ten);
          Token.Let,
          Token.Identifier("result"),
          Token.Equal,
          Token.Identifier("add"),
          Token.LParen,
          Token.Identifier("five"),
          Token.Comma,
          Token.Identifier("ten"),
          Token.RParen,
          Token.Semicolon
        )
      )
    }
  }
}
